facebook/OpenBIC
# Artemis: cb: Support switch heartbeat for DVT

# Description
- Support to get switch heartbeat to judge whether the switch is ready.
  - Enable PWM/TACH to get Switch heartbeat RPM.
- Add patch to avoid utilizeing the busy-wait method in tach_aspeed_sample_fetch function.

# Motivation
- Support to get switch heartbeat to judge whether the switch is ready.

# Test Plan
- Build code: Pass
- Get Switch heartbeat rpm: Pass
- Get Switch temperature sensor reading: Pass
- Check switch heartbeat work on Pre-DVT/EVT2 board: Pass

# Log
- Get Switch temperature sensor reading root@bmc-oob:~# power-util mb status
Power status for fru 1 : OFF
root@bmc-oob:~# sensor-util cb | grep CB_PEX
CB_PEX0_TEMP_C               (0x5) : NA | (na)
CB_PEX1_TEMP_C               (0x6) : NA | (na)
root@bmc-oob:~# fw-util cb --version pesw0; fw-util cb --version pesw1
CB PESW0 Version: NA
CB PESW1 Version: NA
root@bmc-oob:~# power-util mb on
Powering fru 1 to ON state...
root@bmc-oob:~# power-util mb status
Power status for fru 1 : ON
root@bmc-oob:~# sensor-util cb | grep CB_PEX
CB_PEX0_TEMP_C               (0x5) :   60.90 C     | (ok)
CB_PEX1_TEMP_C               (0x6) :   57.81 C     | (ok)
root@bmc-oob:~# fw-util cb --version pesw0; fw-util cb --version pesw1
CB PESW0 Version: 53.11.00.09
CB PESW1 Version: 53.11.01.09

- Check switch heartbeat work on Pre-DVT board
  - MB DC off uart:~$ kernel stacks
    0x4e308 mctptx_01_0a_40 (real size 2048):       unused 1140     usage 908 / 2048 (44 %)
    0x4e240 mctprx_01_0a_40 (real size 4096):       unused 1724     usage 2372 / 4096 (57 %)
    0x83be0 IPMI_thread (real size 4096):   unused 3252     usage 844 / 4096 (20 %)
    0x85f00 sensor_poll (real size 2048):   unused 980      usage 1068 / 2048 (52 %)
    ......
  - MB DC on
    uart:~$ kernel stacks
    0x82768 sw_heartbeat_thread (real size 512):    unused 268      usage 244 / 512 (47 %)
    0x4e308 mctptx_01_0a_40 (real size 2048):       unused 1140     usage 908 / 2048 (44 %)
    0x4e240 mctprx_01_0a_40 (real size 4096):       unused 1724     usage 2372 / 4096 (57 %)
    0x83be0 IPMI_thread (real size 4096):   unused 3252     usage 844 / 4096 (20 %)
    0x85f00 sensor_poll (real size 2048):   unused 980      usage 1068 / 2048 (52 %)
    ......

- Check switch heartbeat work on EVT2 board
  - MB DC off uart:~$ kernel stacks
    0x4d548 mctptx_01_0a_40 (real size 2048):       unused 1140     usage 908 / 2048 (44 %)
    0x4d480 mctprx_01_0a_40 (real size 4096):       unused 1796     usage 2300 / 4096 (56 %)
    0x83be0 IPMI_thread (real size 4096):   unused 3252     usage 844 / 4096 (20 %)
    0x85f00 sensor_poll (real size 2048):   unused 980      usage 1068 / 2048 (52 %)
    ......

  - MB DC on uart:~$ kernel stacks
    0x4d548 mctptx_01_0a_40 (real size 2048):       unused 1140     usage 908 / 2048 (44 %)
    0x4d480 mctprx_01_0a_40 (real size 4096):       unused 1796     usage 2300 / 4096 (56 %)
    0x83be0 IPMI_thread (real size 4096):   unused 3252     usage 844 / 4096 (20 %)
    0x85f00 sensor_poll (real size 2048):   unused 980      usage 1068 / 2048 (52 %)
    ......
